src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs

```diff
@@ -681,6 +682,24 @@ private static IPackageFile FindFileInPackage(string filePath, IEnumerable<IPack
             return null;
         }
 
+        private void ValidateFilesUnique(IEnumerable<IPackageFile> files)
+        {
+            var seen = new HashSet<string>(StringComparer.Ordinal);
+            var duplicates = new HashSet<string>(StringComparer.Ordinal);
```

Comment:
If we're batching all duplicate files for later exception then `duplicates` doesn't have to be HashSet. Because HashSet means we do extra housekeeping work make sure anything going there is unique which is not necessary, we're already did it.
I believe it's better be just `list` type instead of HashSet.

Reply:
duplicates can occur multiple times. You still want to dedupe the duplicates list, though.